crimson/osd/replicate_backend: add the skipped newly created clone object to the push queue after the clone request completes

xxhdx1985126 opened the pull request:
Fixes: https://tracker.ceph.com/issues/68808
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>




<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

Matan-B commented on src/crimson/osd/ops_executer.h:
```diff
@@ -203,6 +203,7 @@ class OpsExecuter {
       ObjectContext& processed_obc) &&;
   };
   std::unique_ptr<CloningContext> cloning_ctx;
+  ObjectContextRef clone_obc;
```
Can we move clone_obc into CloningContext::clone_obc?

Matan-B commented on src/crimson/osd/ops_executer.cc:
```diff
@@ -900,7 +900,7 @@ std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
   }();
 
-  auto clone_obc = prepare_clone(coid, osd_op_params->at_version);
+  clone_obc = prepare_clone(coid, osd_op_params->at_version);
```
cloning_ctx.clone_obc

Matan-B commented on src/crimson/osd/ops_executer.h:
```diff
@@ -519,7 +520,8 @@ OpsExecuter::flush_changes_n_do_ops_effects(
       std::move(*osd_op_params),
-      std::move(log_entries));
+      std::move(log_entries),
+      std::move(clone_obc));
```
```
...
std::move(log_entries),
cloning_ctx ? std::move(cloning_ctx.clone_obc) : nullopt);
```

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -1142,11 +1145,13 @@ PG::submit_executer_fut PG::submit_executer(
 		  auto&& osd_op_p,
-		  auto&& log_entries) {
+		  auto&& log_entries,
+                  auto&& clone_obc) {
```
std::optional<ObjectContextRef&&> clone_obc) {

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -905,10 +905,12 @@ PG::interruptible_future<
   osd_op_params_t&& osd_op_p,
   std::vector<pg_log_entry_t>&& log_entries)
 {
+  auto _clone_obc = std::move(clone_obc);
```
Why is it needed? Can't we just move clone_obc into submit_transaction later on?

xxhdx1985126 replied:
Ah, yes.

Matan-B commented on src/crimson/osd/pg.h:
```diff
@@ -679,6 +679,7 @@ class PG : public boost::intrusive_ref_counter<
     std::tuple<interruptible_future<>, interruptible_future<>>>
   submit_transaction(
     ObjectContextRef&& obc,
+    ObjectContextRef&& clone_obc,
```
Perhaps created_clone_obc / new_clone_obc would be easier to understand?

Matan-B commented on src/crimson/osd/pg_backend.h:
```diff
@@ -417,7 +417,8 @@ class PGBackend
 		     epoch_t min_epoch, epoch_t max_epoch,
-		     std::vector<pg_log_entry_t>&& log_entries) = 0;
+		     std::vector<pg_log_entry_t>&& log_entries,
+		     crimson::osd::ObjectContextRef clone_obc) = 0;
```
ditto

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -905,10 +905,12 @@ PG::interruptible_future<
              PG::interruptible_future<>>>
 PG::submit_transaction(
   ObjectContextRef&& obc,
+  ObjectContextRef&& clone_obc,
```
std::optional <ObjectContextRef&&> created_clone_obc

Matan-B commented on src/crimson/osd/replicated_backend.cc:
```diff
@@ -42,7 +42,8 @@ ReplicatedBackend::submit_transaction(const std::set<pg_shard_t>& pg_shards,
                                       epoch_t min_epoch, epoch_t map_epoch,
-				      std::vector<pg_log_entry_t>&& logv)
+				      std::vector<pg_log_entry_t>&& logv,
+				      crimson::osd::ObjectContextRef clone_obc)
```
ditto

Matan-B reviewed (commented):
This path should be only relevant for backfill osds, right?
I think that passing the (optional) created clone_obc into ReplicatedBackend::submit_transaction unconditionally is quite confusing. Do you know how Classic handles such scenario?

Matan-B commented on src/crimson/osd/osd_operations/snaptrim_event.cc:
```diff
@@ -428,6 +428,7 @@ SnapTrimObjSubEvent::process_and_submit(ObjectContextRef head_obc,
 
   auto [submitted, all_completed] = co_await pg->submit_transaction(
 	  std::move(clone_obc),
+	  nullptr,
```
This is confusing and might be error prune. submit_transaction expects obc and clone_obc, we pass the clone_obc as obc while clone_obc is nullptr. Renaming as suggested earlier might help

Matan-B commented:
Is this DNM until https://github.com/ceph/ceph/pull/59615 is resolved?

xxhdx1985126 commented:
> Is this DNM until #59615 is resolved?

Sorry, I forgot to repush this PR, I've just modified the code to fix a bug in it and follow your suggestions. Please take a look, again. Thanks.

athanatos commented:
I'm confused.  After acquiring the lock on the head, the head must have been pushed to the peer and thus head < last_backfill_started.  Shouldn't the op have been sent to the replica?

athanatos commented:
I don't think adding the clone to the push queue works, btw.  If the head was fully pushed, last_backfill could already be at head past clone.  It needs to be the case that IO processed prior to backfill reaching head skips the replica, and IO processed after that point waits for head to be recovered and then is sent to all replicas.  Subsequently recovering the clone is unlikely to work.

xxhdx1985126 commented:
> I'm confused. After acquiring the lock on the head, the head must have been pushed to the peer and thus head < last_backfill_started. Shouldn't the op have been sent to the replica creating the clone? If not, that's probably what needs to be fixed.

I think if the head is not added to the missing set, the lock on it can be acquired when the pg is under backfilling. The object is added to the missing set when it is added to the push queue in backfill.

In crimson, it is possible that the head is not in the push queue at the time that the pg checks whether it needs to wait for the head to be backfilled, but it does when the replicated backend checks "should_send_op()", which I think is the cause of this issue.

In classic osds, the above case won't happen.

xxhdx1985126 commented:
> head was fully pushed

In this PR, I implemented the push of the newly created clone as a standalone one which doesn't relate to the backfill procedure. So it won't interfere with the ongoing backfill. Am I right? Thanks.

xxhdx1985126 commented:
> > I don't think adding the clone to the push queue works, btw. If the head was fully pushed, last_backfill could already be at head past clone. It needs to be the case that IO processed prior to backfill reaching head skips the replica, and IO processed after that point waits for head to be recovered and then is sent to all replicas. Subsequently recovering the clone is unlikely to work.
> 
> ~In this PR, I implemented the push of the newly created clone as a standalone one which doesn't relate to the backfill procedure. So it won't interfere with the ongoing backfill. Am I right? Thanks.~ Ah, I see your point, even if the clone can be pushed after the IO is done, we can't guarantee this works if the push target goes down before push is done becase `last_backfill` has already gone passed it, right?

@athanatos @Matan-B  Hi, after a second thought. I think it might not be necessary to make sure the consistency between the results of the checks on whether to recover the object and that of whether to send ops to replica osds.

It seems that:
1. if the checks on whether to send the clone op to replica osds fail, `PeeringState::last_backfill` mustn't have been passed the head object as it hasn't been pushed;
2. on the other hand, the newly created clone object's `hobject_t` must be the nearest oid lower than the head object.

So, based on the above two invariant, we can add the newly created object to `BackfillState::progress_tracker` after the io is done and before the obc lock is released by the io request, which would make sure that `PeeringState::last_backfill` don't pass the newly created clone before it's fully backfilled. Am I right?

I've just updated this PR to add the new clone to `BackfillState::progress_tracker`. Please take a look, thanks.

Matan-B commented on src/crimson/osd/osd_operations/snaptrim_event.cc:
```diff
@@ -435,6 +435,7 @@ SnapTrimObjSubEvent::process_and_submit(ObjectContextRef head_obc,
 
   auto [submitted, all_completed] = co_await pg->submit_transaction(
 	  std::move(clone_obc),
+	  ObjectContextRef{},
```
nullptr

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -907,15 +907,30 @@ void PG::mutate_object(
+{
+  assert(recovery_handler);
+  assert(recovery_handler->backfill_state);
+  auto backfill_state = recovery_handler->backfill_state.get();
```
assert(backfill_state)

xxhdx1985126 replied:
> assert(backfill_state)

I thought the line 916 `ssert(recovery_handler->backfill_state);` has the same effect, am I right?

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -907,15 +907,30 @@ void PG::mutate_object(
+  osd_op_params_t&& opp,
   std::vector<pg_log_entry_t>&& log_entries)
 {
+  auto _new_clone = std::move(new_clone);
+  auto osd_op_p = std::move(opp);
```
Isn't this redundant?

xxhdx1985126 replied:
Yeah, will remove it.

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -907,15 +907,30 @@ void PG::mutate_object(
 {
+  auto _new_clone = std::move(new_clone);
+  auto osd_op_p = std::move(opp);
+  auto at_version = osd_op_p.at_version;
```
Can you rebase on top of https://github.com/ceph/ceph/pull/60596 here? As the fix here is not related to the commit.  I'll try to merge 60596 soon.

xxhdx1985126 replied:
Yeah, will do.

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -936,8 +951,32 @@ PG::submit_transaction(
+      if (_new_clone && is_missing_on_peer(pg_shard, hoid)) {
+        // The head is in the push queue but hasn't been pushed,
+        // we need to also add the newly created clone to the push queue;
+        // otherwise, the clone has no chance of being pushed although the
+        // head does.
```
nit
```
        // The head is in the push queue but hasn't been pushed yet.
        // We need to ensure that the newly created clone will be 
        // pushed as well, otherwise we might skip it.
        // See: https://tracker.ceph.com/issues/68808
```

Matan-B commented on src/crimson/osd/ops_executer.h:
```diff
@@ -519,7 +520,10 @@ OpsExecuter::flush_changes_n_do_ops_effects(
+      std::move(log_entries),
+      cloning_ctx
+	? std::move(cloning_ctx->clone_obc)
+	: ObjectContextRef{});
```
nullptr should work, right?

xxhdx1985126 replied:
Yeah

Matan-B commented on src/crimson/osd/pg.cc:
```diff
@@ -936,8 +951,32 @@ PG::submit_transaction(
     projected_log.add(entry);
   }
 
+  std::vector<pg_shard_t> shards_to_send_txn;
```
What do you think about:

`using pg_shard_should_send = std::pair<pg_shard_t, bool>`

`std::vector<pg_shard_should_send> shards_to_send_logs;`

This should make traversing easier later on

xxhdx1985126 replied:
Yeah, will do.

Matan-B commented on src/crimson/osd/replicated_backend.cc:
```diff
@@ -61,36 +102,25 @@ ReplicatedBackend::submit_transaction(const std::set<pg_shard_t>& pg_shards,
-	shard_services.send_to_osd(
-	  pg_shard.osd, std::move(m), map_epoch));
-    }
+  for (auto &pg_shard : shards_to_send_ops) {
```
Continuing previous comment:
`for (auto &[pg_shard, should_send] : shards_to_send_ops)`

Matan-B reviewed (commented):
>So, based on the above two invariant, we can add the newly created object to BackfillState::progress_tracker after the io is done and before the obc lock is released by the io request, which would make sure that PeeringState::last_backfill don't pass the newly created clone before it's fully backfilled. Am I right?

That sound correct to me, @athanatos do you agree?

xxhdx1985126 commented:
jenkins test api

xxhdx1985126 commented:
Sorry, I got the branch wrong.

xxhdx1985126 commented:
@Matan-B All done, please take a look, again. Thanks.

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -936,8 +950,30 @@ PG::submit_transaction(
+        // The head is in the push queue but hasn't been pushed yet.
+        // We need to ensure that the newly created clone will be 
+        // pushed as well, otherwise we might skip it.
+        // See: https://tracker.ceph.com/issues/68808
```
For a client write to get here, either:
1. head <= last_backfill_started and completed the push or
2. head > last_backfill_started

athanatos replied:
I think the difference in behavior between classic and crimson here isn't the classic PG lock, it's that Enqueueing::Enqueueing doesn't take a read lock on the obc before updating last_backfill_started.  Classic, in PrimaryLogPG::recover_backfill, attempts to take a recovery read lock on the object to recover before removing it from the intervals and updating last_backfill_started:

```
	if (obc->get_recovery_read()) {
          ...
      	} else {
	  *work_started = true;
	  dout(20) << "backfill blocking on " << backfill_info.begin
		   << "; could not get rw_manager lock" << dendl;
	  break;
	}
```

By contrast, crimson's Enqueueing::Enqueueing implementation adds the object to the recovery queue and updates last_backfill_started, but does not take a read lock until it actually starts pushing the object some time later.

Note, classic's PG lock here is irrelevant.  The important part is that once the object is <= last_backfill_started, writes will block on the obc lock until the push is complete.  In crimson, however, a client write can begin mutating the object in question before the push actually occurs leading to the race described here.

Adding the clone to the push queue sort seems to solve the problem in that it and the head object will eventually be pushed.  My main reservation with this approach is that this PR does a lot of work to tolerate not sending an IO to a replica for <obc> after <obc> <= last_backfill_started.  I'd rather that case simply not be possible.  However, classic's maintenance of last_backfill_started has some drawbacks:
1. backfill needs a special obc read lock interface
2. the object cannot serve writes on the object between when the push is enqueued and when the push completes

This approach does avoid those problems at the expense of some complexity, and I can't think of any other consequences to allowing repops for <obc> after <obc> <= last_backfill_started to be skipped on a replica.  I think I'm convinced that this should work.

athanatos replied:
TLDR: this approach seems good

xxhdx1985126 replied:
> I think the difference in behavior between classic and crimson here isn't the classic PG lock, it's that Enqueueing::Enqueueing doesn't take a read lock on the obc before updating last_backfill_started. Classic, in PrimaryLogPG::recover_backfill, attempts to take a recovery read lock on the object to recover before removing it from the intervals and updating last_backfill_started:
> 
> ```
> 	if (obc->get_recovery_read()) {
>           ...
>       	} else {
> 	  *work_started = true;
> 	  dout(20) << "backfill blocking on " << backfill_info.begin
> 		   << "; could not get rw_manager lock" << dendl;
> 	  break;
> 	}
> ```
> 
> By contrast, crimson's Enqueueing::Enqueueing implementation adds the object to the recovery queue and updates last_backfill_started, but does not take a read lock until it actually starts pushing the object some time later.
> 
> Note, classic's PG lock here is irrelevant. The important part is that once the object is <= last_backfill_started, writes will block on the obc lock until the push is complete. In crimson, however, a client write can begin mutating the object in question before the push actually occurs leading to the race described here.
> 
> Adding the clone to the push queue seems to solve the problem in that it and the head object will eventually be pushed and last_backfill shouldn't be advanced until both the clone and head are pushed. My main reservation with this approach is that this PR does a lot of work to tolerate not sending an IO to a replica for after <= last_backfill_started. I'd rather that case simply not be possible. However, classic's maintenance of last_backfill_started has some drawbacks:
> 
> 1. backfill needs a special obc read lock interface
> 2. the object cannot serve writes on the object between when the push is enqueued and when the push completes
> 
> This approach does avoid those problems at the expense of some complexity, and I can't think of any other consequences to allowing repops for after <= last_backfill_started to be skipped on a replica. I think I'm convinced that this should work.

@athanatos @Matan-B Hi, sam and matan.

It seems that we have only considered the cases in which only "pushes" are involved. If "deletes" are involved, the current backfill machinery seems to have problem. For example, the primary osd needs to do backfill removal on the backfill targets and a new client request creates the exact object that's to be "backfill-removed", it seems that there's no guarantee in crimson that the backfill removal won't happen after that client request on the backfill targets. Am I right?

xxhdx1985126 replied:
> > I think the difference in behavior between classic and crimson here isn't the classic PG lock, it's that Enqueueing::Enqueueing doesn't take a read lock on the obc before updating last_backfill_started. Classic, in PrimaryLogPG::recover_backfill, attempts to take a recovery read lock on the object to recover before removing it from the intervals and updating last_backfill_started:
> > ```
> > 	if (obc->get_recovery_read()) {
> >           ...
> >       	} else {
> > 	  *work_started = true;
> > 	  dout(20) << "backfill blocking on " << backfill_info.begin
> > 		   << "; could not get rw_manager lock" << dendl;
> > 	  break;
> > 	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > By contrast, crimson's Enqueueing::Enqueueing implementation adds the object to the recovery queue and updates last_backfill_started, but does not take a read lock until it actually starts pushing the object some time later.
> > Note, classic's PG lock here is irrelevant. The important part is that once the object is <= last_backfill_started, writes will block on the obc lock until the push is complete. In crimson, however, a client write can begin mutating the object in question before the push actually occurs leading to the race described here.
> > Adding the clone to the push queue seems to solve the problem in that it and the head object will eventually be pushed and last_backfill shouldn't be advanced until both the clone and head are pushed. My main reservation with this approach is that this PR does a lot of work to tolerate not sending an IO to a replica for after <= last_backfill_started. I'd rather that case simply not be possible. However, classic's maintenance of last_backfill_started has some drawbacks:
> > 
> > 1. backfill needs a special obc read lock interface
> > 2. the object cannot serve writes on the object between when the push is enqueued and when the push completes
> > 
> > This approach does avoid those problems at the expense of some complexity, and I can't think of any other consequences to allowing repops for after <= last_backfill_started to be skipped on a replica. I think I'm convinced that this should work.
> 
> @athanatos @Matan-B Hi, sam and matan.
> 
> It seems that we have only considered the cases in which only "pushes" are involved. If "deletes" are involved, the current backfill machinery seems to have problem. For example, the primary osd needs to do backfill removal on the backfill targets and a new client request creates the exact object that's to be "backfill-removed", it seems that there's no guarantee in crimson that the backfill removal always happens before that client request on the backfill targets. Am I right?

Please ignore this, I was wrong.

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -936,8 +950,30 @@ PG::submit_transaction(
+      shards.emplace_back(pg_shard, true);
+    } else {
+      shards.emplace_back(pg_shard, false);
+      if (_new_clone && is_missing_on_peer(pg_shard, hoid)) {
```
Can we just pass _new_clone into PGBackend::submit_transaction and have it be responsible for figuring out whether it needs to be sent?  I'd prefer that PG::submit_transaction not be determining which shards get the transaction.

xxhdx1985126 replied:
Yeah, will do

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -948,12 +984,19 @@ PG::submit_transaction(
+        enqueue_push_for_backfill(
+          _new_clone->obs.oi.soid,
+          _new_clone->obs.oi.version,
+          to_push_clone);
```
This should be in PGBackend as well.

xxhdx1985126 replied:
Ok.

athanatos commented on src/crimson/osd/pg.cc:
```diff
@@ -907,15 +907,28 @@ void PG::mutate_object(
+  assert(recovery_handler);
+  assert(recovery_handler->backfill_state);
+  auto backfill_state = recovery_handler->backfill_state.get();
+  backfill_state->enqueue_standalone_push(obj, v, peers);
```
PGBackend should probably have a reference to backfill_state/recovery_handler.  Rather than enqueue_standalone_push, how about notify_object_created()?  I'd prefer that PG/PGBackend be notifying the backend of an object being created rather than specifying that a push needs to happen.

xxhdx1985126 commented:
@athanatos @Matan-B All done, please take a look, again. Thanks.

Matan-B commented on src/crimson/osd/replicated_backend.cc:
```diff
@@ -73,18 +74,21 @@ MURef<MOSDRepOp> ReplicatedBackend::new_repop_msg(
   auto log_entries = std::move(logv);
   auto txn = std::move(t);
   auto osd_op_p = std::move(opp);
+  auto _new_clone = std::move(new_clone);
```
nit, redundant

Matan-B reviewed (approved):
lgtm!

Matan-B commented:
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-nov-17

Matan-B commented:
https://pulpito.ceph.com/matan-2024-11-18_09:42:42-crimson-rados-wip-matanb-crimson-testing-nov-17-distro-crimson-smithi/
Rerun:
https://pulpito.ceph.com/matan-2024-11-18_13:14:58-crimson-rados-wip-matanb-crimson-testing-nov-17-distro-crimson-smithi/

Failures:
https://tracker.ceph.com/issues/67791
https://tracker.ceph.com/issues/66606

Matan-B commented:
jenkins test submodules

Matan-B commented:
jenkins test submodules

cyx1231st commented on src/crimson/osd/ops_executer.h:
```diff
@@ -519,14 +521,15 @@ OpsExecuter::flush_changes_n_do_ops_effects(
-      std::move(log_entries));
+      std::move(log_entries),
+      cloning_ctx
+	? std::move(cloning_ctx->clone_obc)
```
I see failure in the latest main:
```
crimson/osd/ops_executer.h:526:27: runtime error: member access within null pointer of type 'struct CloningContext'
```

To reproduce:
```
$MGR=0 MON=1 OSD=1 MDS=0 RGW=0 ceph/src/vstart.sh -n -x --without-dashboard --crimson --nodaemon --redirect-output --seastore
$./bin/ceph osd pool create test-pool 32 32 && ./bin/ceph osd pool set test-pool size 1 --yes-i-really-mean-it && ./bin/rbd create test-pool/test-img --size 50G --image-format=2 --rbd_default_features=3
```

xxhdx1985126 replied:
Ok, will get into it.

xxhdx1985126 replied:
@cyx1231st I can't reproduce this error locally, you are running the main branch, right?

On the other hand, the error looks odd, the codes around `ops_executer.h:526` are as follows:
https://github.com/ceph/ceph/blob/78b4fdf4f580e37367798d07298a89266f100d0a/src/crimson/osd/ops_executer.h#L520-L528

It seems that the access to `cloning_ctx->clone_obc` is protected by the check `cloning_ctx`, which really confuses me.

cyx1231st replied:
see https://github.com/ceph/ceph/pull/60913 or https://github.com/ceph/ceph/pull/60914
